File: algo/CountGoodNodesinBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int goodNodes(TreeNode* root) {
        dfs(root, root->val);
        return res;
    }
    
private:
    int res;
    void dfs(TreeNode* root, int currentMax) {
        if (root == NULL) {
            return;
        }
        if (root->val >= currentMax) {
            res++;
        }
        dfs(root->left, max(currentMax, root->val));
        dfs(root->right, max(currentMax, root->val));
    }
};

File: algo/CountGoodNodesinBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int res;
    public int goodNodes(TreeNode root) {
        res = 0;
        dfs(root, root.val);
        return res;
    }
    

    private void dfs(TreeNode root, int currentMax) {
        if (root == null) {
            return;
        }
        if (root.val >= currentMax) {
            res++;
        }
        dfs(root.left, Math.max(currentMax, root.val));
        dfs(root.right, Math.max(currentMax, root.val));
    }
}